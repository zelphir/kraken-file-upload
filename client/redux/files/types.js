export const GET_FILES_START = 'GET_FILES_START'
export const GET_FILES_SUCCESS = 'GET_FILES_SUCCESS'
export const GET_FILES_FAIL = 'GET_FILES_FAIL'
