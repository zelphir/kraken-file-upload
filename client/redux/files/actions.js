import axios from 'axios'
import * as types from './types'
import { apiUrl } from '../../consts'

export const api = axios.create({ baseURL: apiUrl })

const getFilesStart = () => ({ type: types.GET_FILES_START })
const getFilesSuccess = files => ({ type: types.GET_FILES_SUCCESS, payload: files })
const getFilesFail = error => ({ type: types.GET_FILES_FAIL, error })

export const getFiles = () => async dispatch => {
  dispatch(getFilesStart())
  try {
    const response = await api.get('/files')
    dispatch(getFilesSuccess(response.data))
  } catch (err) {
    dispatch(getFilesFail(err))
  }
}

const deleteFileStart = () => ({ type: types.DELETE_FILE_START })
const deleteFileSuccess = filename => ({ type: types.DELETE_FILE_SUCCESS, payload: filename })
const deleteFileFail = error => ({ type: types.DELETE_FILE_FAIL, error })

export const deleteFile = filename => async dispatch => {
  dispatch(deleteFileStart())
  if (!filename) return dispatch(deleteFileFail(new Error('No filename!')))

  try {
    await api.delete(`/files/${filename}`)
    dispatch(deleteFileSuccess(filename))
  } catch (err) {
    dispatch(deleteFileFail(err))
  }
}

const uploadFileStart = () => ({ type: types.UPLOAD_FILE_START })
const uploadFileSuccess = file => ({ type: types.UPLOAD_FILE_SUCCESS, payload: file })
const uploadFileFail = error => ({ type: types.UPLOAD_FILE_FAIL, error })

export const uploadFile = file => async dispatch => {
  dispatch(uploadFileStart())

  try {
    const response = await api.post('/upload', file)
    dispatch(uploadFileSuccess(response.data))
  } catch (err) {
    dispatch(uploadFileFail(err))
  }
}
