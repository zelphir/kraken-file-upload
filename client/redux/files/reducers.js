import * as types from './types'

export const initialState = {
  data: [],
  isLoading: true,
  error: null,
  query: ''
}

const files = (state = initialState, action) => {
  switch (action.type) {
    case types.GET_FILES_START:
    case types.UPLOAD_FILE_START:
    case types.DELETE_FILE_START:
      return {
        ...state,
        isLoading: true
      }
    case types.GET_FILES_SUCCESS:
      return {
        data: action.payload,
        isLoading: false,
        error: null
      }
    case types.UPLOAD_FILE_SUCCESS:
      return {
        data: [action.payload, ...state.data],
        isLoading: false,
        error: null
      }
    case types.DELETE_FILE_SUCCESS:
      return {
        data: state.data.filter(item => item.filename !== action.payload),
        isLoading: false,
        error: null
      }
    case types.GET_FILES_FAIL:
    case types.UPLOAD_FILE_FAIL:
    case types.DELETE_FILE_FAIL:
      return {
        ...state,
        isLoading: false,
        error: action.error.message
      }
    case types.FILTER_LIST:
      return {
        ...state,
        query: action.payload
        // data: state.data.filter(item => item.filename.includes(action.payload)),
        // isLoading: false,
        // error: null
      }
    default:
      return state
  }
}

export default files
