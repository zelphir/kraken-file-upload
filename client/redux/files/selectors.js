export const getFiles = state => state.files.data
export const isLoading = state => state.files.isLoading
export const getError = state => state.files.error
