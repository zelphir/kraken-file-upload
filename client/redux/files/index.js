import reducer from './reducers'
import * as actions from './actions'
import * as selectors from './selectors'
import * as types from './types'

export { actions, types, selectors }
export default reducer
