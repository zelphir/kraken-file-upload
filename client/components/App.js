import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Callout, Spinner } from '@blueprintjs/core'
import { actions, selectors } from '../redux/files'
import Nav from './Nav'
import List from './List'

export class App extends React.PureComponent {
  static propTypes = {
    isLoading: PropTypes.bool.isRequired,
    data: PropTypes.array.isRequired,
    error: PropTypes.string,
    getFiles: PropTypes.func.isRequired
  }

  componentDidMount() {
    this.props.getFiles()
  }

  render() {
    const { data, isLoading, error } = this.props

    if (error)
      return (
        <div className="spacer">
          <Callout title="Ops, something went wrong!" intent="danger">
            <strong>{error}:</strong> Please try again refreshing the page.
          </Callout>
        </div>
      )

    return isLoading ? (
      <Spinner className="spinner" />
    ) : (
      <React.Fragment>
        <Nav hasData={!!data.length} />
        <List data={data} />
      </React.Fragment>
    )
  }
}

export default connect(
  state => ({
    data: selectors.getFiles(state),
    isLoading: selectors.isLoading(state),
    error: selectors.getError(state)
  }),
  { getFiles: actions.getFiles }
)(App)
