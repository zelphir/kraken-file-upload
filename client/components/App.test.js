import React from 'react'
import { shallow } from 'enzyme'
import { App } from './App'
import { initialState } from '../redux/files/reducers'

function setup(nextProps = {}) {
  const props = {
    getFiles: jest.fn(),
    ...initialState,
    ...nextProps
  }
  const wrapper = shallow(<App {...props} />)

  return {
    props,
    wrapper
  }
}

describe('<App />', () => {
  it('should match the snapshot', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })

  it('should call getFiles() in componentDidMount', () => {
    const { props } = setup()
    expect(props.getFiles.mock.calls.length).toBe(1)
  })

  it('should render the loading message on mount', () => {
    const { wrapper } = setup()
    expect(wrapper.find('Spinner')).toHaveLength(1)
  })

  it('should render the files list after fetching data', () => {
    const file = {
      id: 1,
      filename: 'myfilename.pdf',
      originalName: 'aaaa-bbbb-cccc-dddd.pdf',
      size: 0.05
    }
    const { wrapper } = setup({ isLoading: false, data: [file] })
    expect(wrapper.find('List')).toHaveLength(1)
  })

  it('should render the error message on error', () => {
    const { wrapper } = setup({ error: ':-( Ops, something went wrong! Try again.' })
    expect(wrapper.find('Callout')).toHaveLength(1)
  })
})
