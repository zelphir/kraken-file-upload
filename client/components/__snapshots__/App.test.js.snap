// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match the snapshot 1`] = `
<Blueprint2.Spinner
  className="spinner"
/>
`;
